utils/blockchain.js

```diff
@@ -55,3 +63,51 @@ export async function sendDAOToken(library, toAddress, amount) {
 
   return tx;
 }
+
+export const estimateNeededStakeForNewBooking = ({
+  bookedDates,
+  bookingYear,
+  totalBookingTokenCost,
+}) => {
+  // given the bookedDate is an array of arrays, where
+  // bookedDates[i] = [status, year, dayOfYear, price (BigNumber), timestamp(BigNumber)]
+  // 1. we construct a Map - stakedByYear, where keys are years and values are the total price of bookings for that year in BigNumber
+  // 2. we find how much tokens user spent already in the current year - currentYearStake
+  // 3. we find the max stake among all years - maxStake
+  // 4. we calculate how much tokens user needs to stake for the new booking
+  // 5. we return either 0 if user already spent more than the max stake in the current year,
+  // or the difference between bookingCost, maxStake and currentYearStake
+
+  if (!bookedDates || !bookingYear || !totalBookingTokenCost) {
+    console.error('Missing parameters in estimateNeededStakeForNewBooking');
+    return null;
+  }
+
+  const stakedByYear = bookedDates.reduce((acc, curr) => {
+    const year = curr[1];
+    const price = curr[3];
+    if (acc.has(year)) {
+      acc.set(year, acc.get(year).add(price));
+    } else {
+      acc.set(year, price);
+    }
+    return acc;
+  }, new Map());
+
+  const currentYearStake = formatBigNumberForDisplay(
+    stakedByYear.get(bookingYear),
+    18,
+  );
+
+  const maxStakeBigNum = [...stakedByYear.values()].reduce((acc, curr) => {
+    return acc.gt(curr) ? acc : curr;
+  }, BigNumber.from(0));
+  const maxStake = formatBigNumberForDisplay(maxStakeBigNum, 18);
+
+  const tokensToStake = Math.max(
+    totalBookingTokenCost - Math.max(maxStake, currentYearStake),
+    0,
+  );
+
+  return tokensToStake;
```

Comment:
@juliencharrel I rewrote your function, could you review please? First, I tried to make it more readable in case somebody else would need to use it (or even me in 2-3 months), Secondly, bookedDates is already flat in my case (intentionally), so direct copy/paste from your code doesn't work. 

I also want to have a unit test for all the helpers and this one in particular, lets connect for 10-15 min to defined edge cases and just set some examples?

Reply:
created a ticket for that https://github.com/closerdao/closer-ui/issues/49